<!DOCTYPE html>

<html>
<head>
<meta content="text/html; charset=UTF-8" http-equiv="Content-Type">

<title>module Bio - Pileup</title>

<link type="text/css" media="screen" href="./rdoc.css" rel="stylesheet">

<script type="text/javascript">
  var rdoc_rel_prefix = "./";
</script>

<script type="text/javascript" charset="utf-8" src="./js/jquery.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/navigation.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/search_index.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/search.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/searcher.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/darkfish.js"></script>


<body id="top" class="module">
<nav id="metadata">
  <nav id="home-section" class="section">
  <h3 class="section-header">
    <a href="./index.html">Home</a>
    <a href="./table_of_contents.html#classes">Classes</a>
    <a href="./table_of_contents.html#methods">Methods</a>
  </h3>
</nav>


  <nav id="search-section" class="section project-section" class="initially-hidden">
  <form action="#" method="get" accept-charset="utf-8">
    <h3 class="section-header">
      <input type="text" name="search" placeholder="Search" id="search-field"
             title="Type to search, Up and Down to navigate, Enter to load">
    </h3>
  </form>

  <ul id="search-results" class="initially-hidden"></ul>
</nav>


  <div id="file-metadata">
    <nav id="file-list-section" class="section">
  <h3 class="section-header">Defined In</h3>
  <ul>
    <li>lib/bio/db/pileup.rb
    <li>lib/bio/db/sam/bam.rb
    <li>lib/bio/db/sam/faidx.rb
    <li>lib/bio/db/sam/library.rb
    <li>lib/bio/db/sam/sam.rb
    <li>lib/bio/db/sam.rb
    <li>lib/bio/db/vcf.rb
  </ul>
</nav>

    
  </div>

  <div id="class-metadata">
    
    
    
    
  </div>

  <div id="project-metadata">
    <nav id="fileindex-section" class="section project-section">
  <h3 class="section-header">Pages</h3>

  <ul>
  
    <li class="file"><a href="./LICENSE_txt.html">LICENSE</a>
  
  </ul>
</nav>

    <nav id="classindex-section" class="section project-section">
  <h3 class="section-header">Class and Module Index</h3>

  <ul class="link-list">
  
    <li><a href="./Bio.html">Bio</a>
  
    <li><a href="./Bio/DB.html">Bio::DB</a>
  
    <li><a href="./Bio/DB/Alignment.html">Bio::DB::Alignment</a>
  
    <li><a href="./Bio/DB/Pileup.html">Bio::DB::Pileup</a>
  
    <li><a href="./Bio/DB/SAM.html">Bio::DB::SAM</a>
  
    <li><a href="./Bio/DB/SAM/Library.html">Bio::DB::SAM::Library</a>
  
    <li><a href="./Bio/DB/SAM/Tools.html">Bio::DB::SAM::Tools</a>
  
    <li><a href="./Bio/DB/SAM/Tools/Bam1CoreT.html">Bio::DB::SAM::Tools::Bam1CoreT</a>
  
    <li><a href="./Bio/DB/SAM/Tools/Bam1T.html">Bio::DB::SAM::Tools::Bam1T</a>
  
    <li><a href="./Bio/DB/SAM/Tools/BamHeaderT.html">Bio::DB::SAM::Tools::BamHeaderT</a>
  
    <li><a href="./Bio/DB/SAM/Tools/BamPileup1T.html">Bio::DB::SAM::Tools::BamPileup1T</a>
  
    <li><a href="./Bio/DB/SAM/Tools/SamfileT.html">Bio::DB::SAM::Tools::SamfileT</a>
  
    <li><a href="./Bio/DB/SAM/Tools/SamfileTX.html">Bio::DB::SAM::Tools::SamfileTX</a>
  
    <li><a href="./Bio/DB/SAMException.html">Bio::DB::SAMException</a>
  
    <li><a href="./Bio/DB/Sam.html">Bio::DB::Sam</a>
  
    <li><a href="./Bio/DB/Tag.html">Bio::DB::Tag</a>
  
    <li><a href="./Bio/DB/Vcf.html">Bio::DB::Vcf</a>
  
    <li><a href="./LibC.html">LibC</a>
  
  </ul>
</nav>

  </div>
</nav>

<div id="documentation">
  <h1 class="module">module Bio</h1>

  <div id="description" class="description">
    
<h1 id="label-Bio%3A%3ADB%3A%3APileup+"><a href="Bio/DB/Pileup.html">Bio::DB::Pileup</a> </h1>

<p>A class representing information in SAMTools pileup format</p>
<dl class="rdoc-list note-list"><dt>Author
<dd>
<p>Dan MacLean (tsl at dan.maclean.ac.uk)</p>
</dd></dl>

<p>Pileup is described at <a
href="http://sourceforge.net/apps/mediawiki/samtools/index.php?title=SAM_FAQ#I_do_not_understand_the_columns_in_the_pileup_output">sourceforge.net/apps/mediawiki/samtools/index.php?title=SAM_FAQ#I_do_not_understand_the_columns_in_the_pileup_output</a>.
Briefly (when you invoke pileup with the -c option):</p>
<ul><li>
<p>1 reference sequence name</p>
</li><li>
<p>2 reference coordinate</p>
</li><li>
<p>(3) reference base, or `*‘ for an indel line</p>
</li><li>
<p>(4) genotype where heterozygotes are encoded in the IUB code: M=A/C, R=A/G,
W=A/T, S=C/G, Y=C/T and K=G/T; indels are indicated by, for example,   +A,
-A   or +CC/-C. There is no difference between */+A or +A/*.</p>
</li><li>
<p>(5) Phred-scaled likelihood that the genotype is wrong, which is also
called `consensus quality’.</p>
</li><li>
<p>(6) Phred-scaled likelihood that the genotype is identical to the
reference, which is also called `SNP quality’. Suppose the reference base
is A and in alignment we see 17 G and 3 A. We will get a low consensus
quality because it is difficult to distinguish an A/G heterozygote from a
G/G homozygote. We will get a high SNP quality, though, because the
evidence of a SNP is very strong.</p>
</li><li>
<p>(7) root mean square (RMS) mapping quality</p>
</li><li>
<p>8 # reads covering the position</p>
</li><li>
<p>9 read bases at a SNP line (check the manual page for more information);
the 1st indel allele otherwise</p>
</li><li>
<p>10 base quality at a SNP line; the 2nd indel allele otherwise</p>
</li><li>
<p>(11) indel line only: # reads directly supporting the 1st indel allele</p>
</li><li>
<p>(12) indel line only: # reads directly supporting the 2nd indel allele</p>
</li><li>
<p>(13) indel line only: # reads supporting a third indel allele</p>
</li></ul>

<p>If pileup is invoked without `-c’, indel lines and columns between 3 and
7 inclusive will not be outputted.</p>

<p>NB mpileup uses the 6 column output format eg
“seq2t151tGtGt36t0t99t12t.….……At:9&lt;;;7=&lt;&lt;&lt;&lt;&lt;”
Pileup provides accessors for all columns (6 or 10 column format) and a few
other useful methods</p>

<p>require ‘rubygems’ require ‘ffi’</p>

<pre class="ruby"><span class="ruby-identifier">require</span> <span class="ruby-string">'bio/db/sam/faidx'</span>
<span class="ruby-identifier">require</span> <span class="ruby-string">'bio/db/sam/sam'</span>
</pre>

<p>require ‘rubygems’ require’ffi’ require ‘bio/db/sam/bam’</p>

<pre>sam.rb

Created by Ricardo Ramirez on 3/25/10.</pre>

<p>require ‘rubygems’ require’ffi’ require ‘bio/db/sam/bam’</p>

<h1 id="label-Vcf+">Vcf </h1>

<p>A class representing information in Variant Call Format Forked from vcfruby
at <a
href="https://github.com/jesserod/vcfruby">github.com/jesserod/vcfruby</a> 
Modified and tests written by Dan MacLean (tsl at dan.maclean.ac.uk) VCF is
described at <a
href="http://www.1000genomes.org/node/101">www.1000genomes.org/node/101</a></p>

  </div><!-- description -->

  
  
  
  <section id="5Buntitled-5D" class="documentation-section">
    

    

    

    

    <!-- Methods -->
    
  </section><!-- 5Buntitled-5D -->

</div><!-- documentation -->


<footer id="validator-badges">
  <p><a href="http://validator.w3.org/check/referer">[Validate]</a>
  <p>Generated by <a href="https://github.com/rdoc/rdoc">RDoc</a> 3.12.
  <p>Generated with the <a href="http://deveiate.org/projects/Darkfish-Rdoc/">Darkfish Rdoc Generator</a> 3.
</footer>

